src/stan/analyze/mcmc/compute_effective_sample_size.hpp

```diff
@@ -1,98 +1,184 @@
+        rho_hat_s(s + 2) = rho_hat_odd;
       }
+    }
 
-      return num_chains * num_draws / (1 + 2 * rho_hat_s.sum());
+    int max_s = s;
+    // this is used in the improved estimate
+    if (rho_hat_even > 0)
+      rho_hat_s(max_s + 1) = rho_hat_even;
+
+    // Geyer's initial monotone sequence
+    for (int s = 1; s <= max_s - 3; s += 2) {
+      if (rho_hat_s(s + 1) + rho_hat_s(s + 2) >
+          rho_hat_s(s - 1) + rho_hat_s(s)) {
+        rho_hat_s(s + 1) = (rho_hat_s(s - 1) + rho_hat_s(s)) / 2;
+        rho_hat_s(s + 2) = rho_hat_s(s + 1);
+      }
     }
 
-    inline
-    double compute_effective_sample_size(std::vector<const double*> draws,
-                                         size_t size) {
-      int num_chains = draws.size();
-      std::vector<size_t> sizes(num_chains, size);
-      return compute_effective_sample_size(draws, sizes);
+    double ess = num_chains * num_draws;
+    // Geyer's truncated estimate
+    // Improved estimate reduces variance in antithetic case
+    double tau_hat = -1 + 2*rho_hat_s.head(max_s).sum() + rho_hat_s(max_s + 1);
+    // Safety check for negative values and
+    // with max ess equal to ess*log10(ess)
+    return ess / std::max(tau_hat, 1 / std::log10(ess));
+  }
+
+  /**
+   * Returns the effective sample size for the specified parameter
+   * across all kept samples.
+   *
+   * See more details in Stan reference manual section "Effective
+   * Sample Size". http://mc-stan.org/users/documentation
+   *
+   * Current implementation assumes chains are all of equal size and
+   * draws are stored in contiguous blocks of memory.  Argument size
+   * will be broadcast to same length as draws.
+   *
+   * @param std::vector stores pointers to arrays of chains
+   * @param size_t size of chains
+   * @return effective sample size for the specified parameter
+   */
+  inline
+  double compute_effective_sample_size(std::vector<const double*> draws,
+                                       size_t size) {
+    int num_chains = draws.size();
+    std::vector<size_t> sizes(num_chains, size);
+    return compute_effective_sample_size(draws, sizes);
+  }
+
+  /**
+   * Returns the effective sample size for the specified parameter
+   * across all kept and split samples.
+   *
+   * See more details in Stan reference manual section "Effective
+   * Sample Size". http://mc-stan.org/users/documentation
+   *
+   * Current implementation assumes chains are all of equal size and
+   * draws are stored in contiguous blocks of memory.
+   *
+   * @param std::vector stores pointers to arrays of chains
+   * @param std::vector stores sizes of chains
+   * @return effective sample size for the specified parameter
+   */
+  inline
+  double compute_split_effective_sample_size(std::vector<const double*> draws,
+                                             std::vector<size_t> sizes) {
+    int num_chains = sizes.size();
+
+    // need to generalize to each jagged draws per chain
```

Comment:
No reason for speculative comments here.  Current implementation is limited to non-ragged chains.